### PR TITLE
feat: new BlockReader implementation for block-as-file

### LINKED ISSUE #349: feat: new BlockReader implementation for block-as-file
## Story

**AS A** Block Node Operator/User
**I WANT** to be able to retrieve blocks written as files
**SO THAT** I will be able to use the blocks

## Tech Notes

- **DEPENDS ON** #281

- See #125 where we have deigned what the functionality of the `block-as-file` format should be.

- The scope of this issue is to be able to accurately, performantly and reliably read files from the file system, that represent any given `block-as-file`.

- Initially, we should be able to read only the raw files that are written to the file system, then, as the file writing logic gets extended where it is being compressed optionally and then also archived, we should extend the reading functionality to support that as well.

### PR DESCRIPTION
**Pull request by ata-nas**
**Description**:

- adding reading logic to complement #281 

**Related issue(s)**:

Fixes #349 

**Notes for reviewer**:

- what would be the best way to parse the block file, having a file input stream and calling readAllBytes does not seem like the best option, there are other (based on the api), but which one would be preferable?

**Checklist**

- [x] Documented (Code comments, README, etc.)
- [x] Tested (unit, integration, etc.)

### CODE REVIEW

File: server/src/main/java/com/hedera/block/server/persistence/storage/read/BlockAsLocalFileReader.java [hunk 1]
```diff
@@ -17,34 +17,65 @@
...
+            return Optional.of(doRead(resolvedBlockPath));
+        } else {
+            return Optional.empty();
+        }
+    }
+
+    private BlockUnparsed doRead(final Path resolvedBlockPath) throws IOException, ParseException {
+        // todo there are other ways of parsing, for example we can pass ReadableSequentialData instead of the fos
+        // because the readAllBytes does not seem like the best idea, what is the correct (or better) way to parse
+        // the block file?
```
Comment (ata-nas) [inline server/src/main/java/com/hedera/block/server/persistence/storage/read/BlockAsLocalFileReader.java:76]: Opinions?

File: server/src/main/java/com/hedera/block/server/persistence/storage/read/BlockAsLocalFileReader.java [hunk 2]
```diff
@@ -17,34 +17,68 @@
...
+    public Optional<BlockUnparsed> read(final long blockNumber) throws IOException, ParseException {
         Preconditions.requireWhole(blockNumber);
-        throw new UnsupportedOperationException("Not implemented yet");
+        final Path resolvedBlockPath = pathResolver.resolvePathToBlock(blockNumber);
+        if (Files.exists(resolvedBlockPath)) {
+            return Optional.of(doRead(resolvedBlockPath));
+        } else {
+            // todo we generally expect the block to be there, so we
+            // return an empty optional here, or do something else?
+            // of course sometimes the block may not even exist yet.
```
Comment (ata-nas) [inline server/src/main/java/com/hedera/block/server/persistence/storage/read/BlockAsLocalFileReader.java:71]: Opinions?

Comment (AlfredoG87) [reply server/src/main/java/com/hedera/block/server/persistence/storage/read/BlockAsLocalFileReader.java:71]: I agree that returning an empty optional is correct.

File: server/src/main/java/com/hedera/block/server/persistence/storage/read/BlockAsLocalFileReader.java [hunk 3]
```diff
@@ -17,34 +17,68 @@
...
+            return Optional.empty();
+        }
+    }
+
+    private BlockUnparsed doRead(final Path resolvedBlockPath) throws IOException, ParseException {
+        // todo there are other ways of parsing, for example we can pass ReadableSequentialData instead of the fis
+        // because the readAllBytes does not seem like the best idea, what is the correct (or better) way to parse
+        // the block file?
+        try (final FileInputStream fis = new FileInputStream(resolvedBlockPath.toFile())) {
+            return BlockUnparsed.PROTOBUF.parse(Bytes.wrap(fis.readAllBytes()));
```
Comment (AlfredoG87) [inline server/src/main/java/com/hedera/block/server/persistence/storage/read/BlockAsLocalFileReader.java:81]: I think that the biggest block files will be around the 10Mb, so I would say we don't need to bother with a Buffered approach for now. But might be good to create a ticket with future tag to keep it in the radar 👍

**Review by AlfredoG87 — APPROVED** [review]
LGTM

File: server/src/main/java/com/hedera/block/server/persistence/storage/read/BlockAsLocalFileReader.java (hunk 1 above)
Comment (jsync-swirlds) [reply server/src/main/java/com/hedera/block/server/persistence/storage/read/BlockAsLocalFileReader.java]: I am assuming, here, that you intend to just wrap the input stream in the readable sequential data and pass that to the parse method instead of calling `readAllBytes`.

I believe that constructing a `ReadableSequentialData` here, instead of reading the full file as bytes before parsing, will save a byte array copy, and should improve memory efficiency for minimal change in complexity.

File: server/src/test/java/com/hedera/block/server/persistence/storage/read/BlockAsLocalFileReaderTest.java [hunk 4]
```diff
@@ -29,25 +51,64 @@
...
      * a given block number.
-     *
-     * @param toRead parameterized, block number
      */
-    @ParameterizedTest
-    @MethodSource("validBlockNumbers")
-    void testSuccessfulBlockRead(final long toRead) {
-        // todo currently throws UnsupportedOperationException
-        assertThatExceptionOfType(UnsupportedOperationException.class).isThrownBy(() -> toTest.read(toRead));
+    @Test
```
Comment (jsync-swirlds) [inline server/src/test/java/com/hedera/block/server/persistence/storage/read/BlockAsLocalFileReaderTest.java:82]: Why remove the parameterized test case source?  This runs fewer scenarios and reduces the number of requirements validated...

File: server/src/main/java/com/hedera/block/server/persistence/storage/read/BlockAsLocalFileReader.java (hunk 2 above)
Comment (jsync-swirlds) [reply server/src/main/java/com/hedera/block/server/persistence/storage/read/BlockAsLocalFileReader.java:71]: Already removed, but I would agree that a missing block file is _normal_ and should be handled without exceptions.

File: server/src/main/java/com/hedera/block/server/persistence/storage/read/BlockAsLocalFileReader.java (hunk 3 above)
Comment (jsync-swirlds) [reply server/src/main/java/com/hedera/block/server/persistence/storage/read/BlockAsLocalFileReader.java:81]: We found in CN that we need to buffer reads because otherwise the parse is _very_ costly as it reads one byte at a time (which causes excessive lock contention in the java internals).

File: server/src/main/java/com/hedera/block/server/persistence/storage/read/BlockAsLocalFileReader.java (hunk 1 above)
Comment (ata-nas) [reply server/src/main/java/com/hedera/block/server/persistence/storage/read/BlockAsLocalFileReader.java]: ✅ Indeed, wrapping the input stream is what I meant. I've addressed that, thanks!

File: server/src/main/java/com/hedera/block/server/persistence/storage/read/BlockAsLocalFileReader.java (hunk 2 above)
Comment (ata-nas) [reply server/src/main/java/com/hedera/block/server/persistence/storage/read/BlockAsLocalFileReader.java]: ✅ Thanks for the clarification @AlfredoG87 @jsync-swirlds

File: server/src/main/java/com/hedera/block/server/persistence/storage/read/BlockAsLocalFileReader.java (hunk 3 above)
Comment (ata-nas) [reply server/src/main/java/com/hedera/block/server/persistence/storage/read/BlockAsLocalFileReader.java]: ✅ Thanks, I did migrate to the readable sequential data.

File: server/src/test/java/com/hedera/block/server/persistence/storage/read/BlockAsLocalFileReaderTest.java (hunk 4 above)
Comment (ata-nas) [reply server/src/test/java/com/hedera/block/server/persistence/storage/read/BlockAsLocalFileReaderTest.java:82]: The proper setup if I have kept the parameterized test was too difficult to do. I do actually write to the fs here, so it requires writing the test data first and then reading it. I'll see how to improve this with more test cases. Also, I will make another test to assert that an empty optional is returned if the block file does not exist (but block number is valid)

**Review by mattp-swirldslabs — APPROVED** [review]
Looks good

File: server/src/test/java/com/hedera/block/server/persistence/storage/read/BlockAsLocalFileReaderTest.java (hunk 4 above)
Comment (jsync-swirlds) [reply server/src/test/java/com/hedera/block/server/persistence/storage/read/BlockAsLocalFileReaderTest.java:82]: Should have been relatively straightforward, mostly the test you have, but the block number is a parameter (and that determines the file name).
It will write a file, but it should also remove that file at the end of the test method, and in between it is basically unchanged...

##### General note
The use of parameterized tests is encouraged for _all_ unit tests; if the unit test is hard to parameterize, it probably needs to be rethought (not always, but in most cases, it can be parameterized with less change than you might expect).

File: server/src/test/java/com/hedera/block/server/persistence/storage/read/BlockAsLocalFileReaderTest.java [hunk 5]
```diff
@@ -29,25 +51,64 @@
...
+                .isExactlyInstanceOf(BlockUnparsed.class)
+                .returns(1L, from(blockUnparsed -> {
+                    try {
+                        return BlockHeader.PROTOBUF
+                                .parse(Objects.requireNonNull(
+                                        blockUnparsed.blockItems().getFirst().blockHeader()))
+                                .number();
+                    } catch (final ParseException e) {
+                        throw new RuntimeException(e);
+                    }
```
Comment (jsync-swirlds) [inline server/src/test/java/com/hedera/block/server/persistence/storage/read/BlockAsLocalFileReaderTest.java:104]: FYI, unit tests generally should not catch exceptions; let the exception be thrown from the test method and the test runner will give you better information in the test report.

File: server/src/test/java/com/hedera/block/server/persistence/storage/read/BlockAsLocalFileReaderTest.java [hunk 6]
```diff
@@ -29,25 +51,64 @@
...
+                }))
+                .extracting(BlockUnparsed::blockItems)
+                .asList()
+                .isNotNull()
+                .isNotEmpty()
+                .hasSize(blockItemUnparsed.size())
+                .containsExactlyElementsOf(blockItemUnparsed);
```
Comment (jsync-swirlds) [inline server/src/test/java/com/hedera/block/server/persistence/storage/read/BlockAsLocalFileReaderTest.java:111]: These two items are almost certainly overly specific and make the test brittle.
Verifying that what is read is _exactly_ what was written is better done in a separate test _by itself_ with its own input to reduce the probability of tests broken in ways that have nothing to do with whether the design is correct.

File: server/src/main/java/com/hedera/block/server/persistence/storage/read/BlockAsLocalFileReader.java [hunk 7]
```diff
@@ -17,34 +17,61 @@
...
-    public Optional<BlockUnparsed> read(final long blockNumber) {
+    public Optional<BlockUnparsed> read(final long blockNumber) throws IOException, ParseException {
         Preconditions.requireWhole(blockNumber);
-        throw new UnsupportedOperationException("Not implemented yet");
+        final Path resolvedBlockPath = pathResolver.resolvePathToBlock(blockNumber);
+        if (Files.exists(resolvedBlockPath)) {
+            return Optional.of(doRead(resolvedBlockPath));
+        } else {
+            return Optional.empty();
+        }
```
Comment (jsync-swirlds) [inline server/src/main/java/com/hedera/block/server/persistence/storage/read/BlockAsLocalFileReader.java:69]: Small note:
A future refactor should push the "find this file" into the resolver, so if the `Path` is not "empty" from the resolver (which probably should return an `Optional`), we are guaranteed that the file exists.  That both supports the resolver testing multiple paths (e.g. for zip archives) and also simplifies this method.

File: server/src/test/java/com/hedera/block/server/persistence/storage/read/BlockAsLocalFileReaderTest.java (hunk 4 above)
Comment (ata-nas) [reply server/src/test/java/com/hedera/block/server/persistence/storage/read/BlockAsLocalFileReaderTest.java]: ✅ I have streamlined this, the parameterized tests are back and I have added another helper method to generate block items for a block with specific number to facilitate this.

File: server/src/test/java/com/hedera/block/server/persistence/storage/read/BlockAsLocalFileReaderTest.java (hunk 5 above)
Comment (ata-nas) [reply server/src/test/java/com/hedera/block/server/persistence/storage/read/BlockAsLocalFileReaderTest.java]: I agree, however the `returns` test requires a `Function` which unfortunately does not throw, and the `parse` method throws an unchecked exception. I do re-throw it immediately however so we do not silence the exception (as it should be the case).

File: server/src/main/java/com/hedera/block/server/persistence/storage/read/BlockAsLocalFileReader.java (hunk 7 above)
Comment (ata-nas) [reply server/src/main/java/com/hedera/block/server/persistence/storage/read/BlockAsLocalFileReader.java:69]: That is a really good point! I will make sure to add that in the upcoming PRs where we will be also using the archive root (we will need this logic there as well) and will migrate such uses of the `if-else` where we ask if the path to a block resolved exists.

File: server/src/test/java/com/hedera/block/server/persistence/storage/read/BlockAsLocalFileReaderTest.java (hunk 6 above)
Comment (ata-nas) [reply server/src/test/java/com/hedera/block/server/persistence/storage/read/BlockAsLocalFileReaderTest.java:140]: ✅ That is a good point, I split the tests. I do agree that the tests should be testing specific aspects of a given method and when there are such clearly defined and visible aspects of a method, like the method relies on something to be present on the filesystem and only then it can read the contents, then these are indeed two different tests, because if one breaks, this would very easily indicate which aspect of the intended functionality was broken.

However for the brittleness, while I do agree with you that it is good to have locally generated data, I have added another utility method to generate the test data so it could be reused. I think that if there is another case, where we need test data generated differently, we should add another method and not tamper with the logic of the already existing utility method that generates our data. For such a complex data generation logic, I do actually think it is better to indeed be in a publicly accessible utility, because otherwise if something changes with the data model, it will be a very big hustle to support.

**Review by AlfredoG87 — APPROVED** [review]
a few nits, looks good!

File: server/src/main/java/com/hedera/block/server/persistence/storage/read/BlockAsLocalFileReader.java (hunk 2 above)
Comment (AlfredoG87) [inline server/src/main/java/com/hedera/block/server/persistence/storage/read/BlockAsLocalFileReader.java]: remove the comments?

File: server/src/main/java/com/hedera/block/server/persistence/storage/read/BlockAsLocalFileReader.java [hunk 8]
```diff
@@ -17,34 +17,68 @@
...
+            // return an empty optional here, or do something else?
+            // of course sometimes the block may not even exist yet.
+            return Optional.empty();
+        }
+    }
+
+    private BlockUnparsed doRead(final Path resolvedBlockPath) throws IOException, ParseException {
+        // todo there are other ways of parsing, for example we can pass ReadableSequentialData instead of the fis
+        // because the readAllBytes does not seem like the best idea, what is the correct (or better) way to parse
+        // the block file?
```
Comment (AlfredoG87) [inline server/src/main/java/com/hedera/block/server/persistence/storage/read/BlockAsLocalFileReader.java]: remove the comments?

Comment (ata-nas) [reply server/src/main/java/com/hedera/block/server/persistence/storage/read/BlockAsLocalFileReader.java]: @AlfredoG87 you seem to be looking at an old version of this file, changes were already pushed before you have submitted your review.

File: server/src/main/java/com/hedera/block/server/persistence/storage/read/BlockAsLocalFileReader.java (hunk 2 above)
Comment (ata-nas) [reply server/src/main/java/com/hedera/block/server/persistence/storage/read/BlockAsLocalFileReader.java]: @AlfredoG87 you seem to be looking at an old version of this file, changes were already pushed before you have submitted your review.

File: server/src/main/java/com/hedera/block/server/persistence/storage/read/BlockAsLocalFileReader.java (hunk 3 above)
Comment (ata-nas) [reply server/src/main/java/com/hedera/block/server/persistence/storage/read/BlockAsLocalFileReader.java]: @AlfredoG87 you seem to be looking at an old version of this file.

**Review by AlfredoG87 — APPROVED** [review]
LGTM
